### PR TITLE
feat: delete timed out CI clusters [DET-3504]

### PR DESCRIPTION
**Pull request by eecsliu**
## Description
Tests in CircleCI time out after six hours, but the stacks that are created are not deleted. This change adds a workflow to CircleCI that deletes all stacks containing the name "nightly", "parallel", or "e2e-gpu" and have existed for more than six hours. This workflow is set to run every six hours.

## Commentary (optional)

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1262,3 +1273,15 @@ workflows:
...
+    triggers:
+      - schedule:
+          cron: "0 5 * * *"
+          filters:
+            branches:
+              only:
+                - master
+    jobs:
+      - cleanup-stacks:
+          context: aw
```
Comment (aaron276h) [inline .circleci/config.yml:1287]: blocking: this should be `aws` and add a new line and the end of the file

File: .circleci/scripts/remove_stacks.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import boto3
+import datetime
+import os
+
+from dateutil.tz import tzutc
+
+
+def check_conditions(stack: dict, target_stacks: dict, time_diff: datetime.timedelta):
```
Comment (aaron276h) [inline .circleci/scripts/remove_stacks.py:8]: non-blocking: could we rename this function to `check_stack_timeout`. Also could we update the function args to describe the types a bit more:

```
def check_stack_timeout(stack: Dict[str, Any], target_stacks: Dict[str, List[Any]], timeout: datetime.timedelta) -> None:
```

File: .circleci/scripts/remove_stacks.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    difference = datetime.timedelta(hours=6)
+
+    targetStacks = {'nightly':[], 'e2e-gpu':[], 'parallel':[]}
+
+    for each in response["Stacks"]:
+        check_conditions(each, targetStacks, difference)
+
+    for k, lists in targetStacks.items():
+        for each in lists:
+            os.system("det-deploy aws down --cluster-id " + each)
```
Comment (aaron276h) [inline .circleci/scripts/remove_stacks.py:28]: Use an f"" for this so: ` os.system(f"det-deploy aws down --cluster-id {each}")`

File: .circleci/scripts/remove_stacks.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    response = client.describe_stacks()
+    difference = datetime.timedelta(hours=6)
+
+    targetStacks = {'nightly':[], 'e2e-gpu':[], 'parallel':[]}
+
+    for each in response["Stacks"]:
+        check_conditions(each, targetStacks, difference)
+
+    for k, lists in targetStacks.items():
+        for each in lists:
```
Comment (aaron276h) [inline .circleci/scripts/remove_stacks.py:27]: ? `each` -> `stack`

File: .circleci/scripts/remove_stacks.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+    for k in target_stacks:
+        if k in stack["StackName"] and current_time - stack["CreationTime"] > time_diff:
+            target_stacks[k].append(stack["StackName"])
+
+
+if __name__ == "__main__":
+    client = boto3.client("cloudformation")
+    response = client.describe_stacks()
+    difference = datetime.timedelta(hours=6)
```
Comment (aaron276h) [inline .circleci/scripts/remove_stacks.py:19]: ? rename this to timeout

**Comment by eecsliu** [comment]
For the slack notifications, would it just be a notification that a deletion happened? Or should it include specifics like the name of the stacks deleted?

**Comment by aaron276h** [comment]
@eecsliu it would be nice to include the stack name(s) in the slack message if possible, but it would also work just to have a slack message alerting that a deletion happened and log the stack name in circle CI logs.

**Comment by eecsliu** [comment]
@aaron276h Do you know if there is a way to add multiple contexts to a workflow? It would be nice to use the slack API in python to send notifications as the deletions are happening. From what I've seen though, it looks like you can't have multiple contexts, so I've been trying to think of another way to keep the list of deleted stacks. 

Possible solution is setting an environment variable with all the stacks listed? Do you have any thoughts?

**Comment by aaron276h** [comment]
@eecsliu Those are both good options. You are right it is not possible to use multiple contexts, but it is possible to add new environment variables to an existing context, so in this case we can use the `SLACK_WEBHOOK` if we need it. Also looking at the [slack circle ci orb](https://circleci.com/orbs/registry/orb/circleci/slack#commands-status) it uses the the `SLACK_WEBHOOK` so it appears to already may be set.

**Pull request closed without merging**